### PR TITLE
	Don't validate selector that is already validated

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #30875

@timothysc

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30976)

### CODE REVIEW

**Comment by gmarek** [comment]
LGTM

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30976#-:-KPXFOqv80dc6Y1EdghW:bb6ywmx)*

File: pkg/labels/selector.go [hunk 1]
```diff
@@ -800,6 +800,22 @@ func SelectorFromSet(ls Set) Selector {
 	return internalSelector(requirements)
 }
 
+// SelectorFromValidatedSet returns a Selector which will match exactly the given Set.
+// A nil and empty Sets are considered equivalent to Everything().
+// It assumes that Set is already validated and doesn't do any validation.
+func SelectorFromValidatedSet(ls Set) Selector {
+	if ls == nil {
```
Comment (lavalamp) [inline pkg/labels/selector.go]: move body into other function, there's no need to have two functions that I can see. This isn't used elsewhere.

File: pkg/labels/labels.go [hunk 2]
```diff
@@ -61,6 +61,13 @@ func (ls Set) AsSelector() Selector {
 	return SelectorFromSet(ls)
 }
 
+// ValidatedAsSelector converts labels into a selector, but
+// assumes that labels are already validated and thus don't
+// preform any validation.
+func (ls Set) ValidatedAsSelector() Selector {
```
Comment (lavalamp) [inline pkg/labels/labels.go]: Better names?

AsSelectorPreValidated
AsSelectorFastPath
UnsafeAsSelector
UnvalidatedAsSelector

File: pkg/labels/labels.go [hunk 3]
```diff
@@ -61,6 +61,13 @@ func (ls Set) AsSelector() Selector {
 	return SelectorFromSet(ls)
 }
 
+// ValidatedAsSelector converts labels into a selector, but
+// assumes that labels are already validated and thus don't
+// preform any validation.
```
Comment (lavalamp) [inline pkg/labels/labels.go]: "we measured and this is significantly faster in codepaths that matter at high scale"

**Comment by lavalamp** [comment]
Naming suggestions, lgtm otherwise.

File: pkg/labels/labels.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/labels/labels.go]: done

File: pkg/labels/labels.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/labels/labels.go]: done

File: pkg/labels/selector.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/labels/selector.go]: For now I'm leaving it here for consistency with "SelectorFromSet". If you feel really strongly about it, I will send another PR with it.

**Comment by wojtek-t** [comment]
Comments applied - self-appling lgtm label based on comments above.

**Comment by wojtek-t** [comment]
@k8s-bot verify this, issue: #IGNORE (Jenkins issue)

**Comment by justinsb** [comment]
@wojtek-t did you see my comment asking why this validation is so expensive?  Perhaps I'm missing something, but if we have a super-expensive regex, fixing that would benefit all code paths: https://github.com/kubernetes/kubernetes/issues/30875#issuecomment-240780402

**Comment by wojtek-t** [comment]
@justinsb - yeah, sorry, forgot to answer;
yes - it's because of some regex, but I didn't have time to debug which exactly regex it is (though I think we have just two of them). I agree that fixing it might help, but this will probably be non-trivial change (though probably super local) and would require much more work (and I didn't have time before 1.4 to do it).

**Comment by justinsb** [comment]
Thanks @wojtek-t and makes sense.  I opened #31124 so we don't lose the optimization opportunity.  I also tagged it `help-wanted` as this seems the sort of thing that could be a nice first contribution for someone to k8s, but past experience says we should not assume anyone will take it up.

**Pull request merged**